### ISSUE TITLE
Strange behaviour of Huber loss function regarding L1 section

### ISSUE THREAD

**Issue by miqlar** [microsoft/LightGBM#2762]
##  Environment info

C++/Python/R version: Python 3
LightGBM version: 2.3.1

## Summary

Huber loss is a mix of MAE (L1) and MSE (L2). So for abs(error)>alpha, huber is a L1 if alpha=1. Illustration below for clarity (MAE in red).

![Screenshot 2020-02-14 at 16 02 39 (2)](https://user-images.githubusercontent.com/8180563/74542772-7a830a00-4f44-11ea-90a4-c57d1625b9cd.png)

Basically - I was getting strange results with huber and decided to test the loss function.

**If alpha is very high, the loss acts as a MSE and the model learns the same. But if alpha=1, for a regression task with high target values (to force using the MAE zone), results diverge a lot from MAE (I would expect them to be similar at least).**

I have also looked into the implementation and haven't seen any reason for this behaviour.

https://github.com/microsoft/LightGBM/blob/1c27a15e42f0076492fcc966b9dbcf9da6042823/src/objective/regression_objective.hpp#L197
https://github.com/microsoft/LightGBM/blob/1c27a15e42f0076492fcc966b9dbcf9da6042823/src/objective/regression_objective.hpp#L292

Is there any reason for this? Is there something I am overlooking?

## Reproducible example

```
import pandas as pd
import lightgbm as lgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import seaborn as sns

df  = pd.read_csv('winequality-red.csv', sep=';')

X = df.drop(columns=['quality'])
y = df.quality*10000 #Make target have high values

model_mae   = lgb.LGBMRegressor(objective='mae')
model_huber_alpha_standard = lgb.LGBMRegressor(objective='huber', alpha=1)

model_mse   = lgb.LGBMRegressor(objective='mse')
model_huber_alpha_big = lgb.LGBMRegressor(objective='huber', alpha=10e10)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

model_mae.fit(X_train, y_train)
model_mse.fit(X_train, y_train)
model_huber_alpha_standard.fit(X_train, y_train)
model_huber_alpha_big.fit(X_train, y_train)

print("MSE:                             {}".format(mean_squared_error(y_test, model_mse.predict(X_test))))
print("Huber with huge alpha - MSE:     {}".format(mean_squared_error(y_test, model_huber_alpha_big.predict(X_test))))
print()
print("MAE:                             {}".format(mean_squared_error(y_test, model_mae.predict(X_test))))
print("Huber with alpha=1 - Pseudo MAE: {}".format(mean_squared_error(y_test, model_huber_alpha_standard.predict(X_test))))

```

Which results in:

```
MSE:                             37480993.22941556
Huber with huge alpha - MSE:     37480993.22941556

MAE:                             39433560.93471763
Huber with alpha=1 - Pseudo MAE: 63570885.04436089
```

Dataset: https://archive.ics.uci.edu/ml/datasets/Wine+Quality

**Comment by guolinke** [comment]
You can check the RenewTreeOutput in mae objective. 
The built in mae objective will reset tree outputs to the medium of residuals.

**Comment by miqlar** [comment]
Hi,

I read about the special treatment for loss functions where the gradient is constant in https://github.com/microsoft/LightGBM/issues/2128. Should't this be applied in this case too for the huber in the L1 section too, then? Thanks!

**Comment by guolinke** [comment]
@miqlar this is not used into huber. 
I think Huber is a smooth version of MAE, and optimizing huber is often for the MAE metric. 
Therefore, I don't think that changing Huber and make it like our specific treatment in L1 is a good idea, otherwise, it is very similar to the built-in MAE objective.
